torch/utils/_pytree.py

```diff
@@ -0,0 +1,150 @@
+from typing import NamedTuple, Callable, Any, Tuple, List, Dict
+
+"""
+Contains utility functions for working with nested python data structures.
+
+A *pytree* is Python nested data structure. It is a tree in the sense that
+nodes are Python collections (e.g., list, tuple, dict) and the leaves are
+Python values. Furthermore, a pytree should not contain reference cycles.
+
+pytrees are useful for working with nested collections of Tensors. For example,
+one can use `tree_map` to map a function over all Tensors inside some nested
+collection of Tensors and `tree_unflatten` to get a flat list of all Tensors
+inside some nested collection. pytrees are helpful for implementing nested
+collection support for PyTorch APIs.
+
+This pytree implementation is not very performant due to Python overhead
+To improve the performance we can move parts of the implementation to C++.
+"""
+
+# A NodeDef holds two callables:
+# - flatten_fn should take the collection and return a flat list of values.
+#   It can also return some context that is used in reconstructing the
+#   collection.
+# - unflatten_fn should take a flat list of values and some context
+#   (returned by flatten_fn). It returns the collection by reconstructing
+#   it from the list and the context.
+Context = Any
+pytree_t = Any
+flatten_fn_t = Callable[[pytree_t], Tuple[List, Context]]
+unflatten_fn_t = Callable[[List, Context], pytree_t]
+NodeDef = NamedTuple('NodeDef', [
+    ('flatten_fn', flatten_fn_t),
+    ('unflatten_fn', unflatten_fn_t),
+])
```

Comment:
protip, you can also write it as:

```
class NodeDef(NamedTuple):
  flatten_fn: flatten_fn_t
  unflatten_fn: unflatten_fn_t
```

Reply:
This makes it look a lot nicer, thanks for pointing that out!